python/lsst/pipe/base/all_dimensions_quantum_graph_builder.py

```diff
@@ -128,262 +127,314 @@ def process_subgraph(self, subgraph: PipelineGraph) -> QuantumGraphSkeleton:
+            )
+        query_cmd: list[str] = []
+        with self.butler.query() as query:
+            query_cmd.append("with butler.query() as query:")
+            query_cmd.append(f"    query = query.join_dimensions({list(tree.all_dimensions.names)})")
+            query = query.join_dimensions(tree.all_dimensions)
+            if constraint_datasets:
+                query_cmd.append(f"    collections = {list(self.input_collections)}")
+            for dataset_type_name in constraint_datasets:
+                query_cmd.append(f"    query = query.join_dataset_search({dataset_type_name!r}, collections)")
+                query = query.join_dataset_search(dataset_type_name, self.input_collections)
+            query_cmd.append(
+                f"    query = query.where({dict(tree.subgraph.data_id.mapping)}, "
+                f"{self.where!r}, bind={self.bind!r})"
+            )
+            query = query.where(tree.subgraph.data_id, self.where, bind=self.bind)
+            self.log.verbose("Querying for data IDs via: %s", "\n".join(query_cmd))
+            # Allow duplicates from common skypix overlaps to make some queries
+            # run faster.
+            query._allow_duplicate_overlaps = True
+            self.log.info("Iterating over query results to associate quanta with datasets.")
+            # Iterate over query results, populating data IDs for datasets,
+            # quanta, and edges.  We populate only the first level of the tree
+            # in the first pass, so we can be done with the query results as
+            # quickly as possible in case that holds a connection/cursor open.
+            n_rows = 0
+            for common_data_id in query.data_ids(tree.all_dimensions):
+                for branch_dimensions, branch in tree.trunk_branches.items():
+                    data_id = common_data_id.subset(branch_dimensions)
+                    branch.data_ids.add(data_id)
+                n_rows += 1
+            if n_rows == 0:
+                # A single multiline log plays better with log aggregators like
+                # Loki.
+                lines = ["Initial data ID query returned no rows, so QuantumGraph will be empty."]
+                try:
+                    lines.extend(query.explain_no_results())
+                finally:
+                    lines.append("To reproduce this query for debugging purposes, run:")
+                    lines.extend(query_cmd)
+                    # If an exception was raised, write a partial.
+                    self.log.error("\n".join(lines))
+                return
+        self.log.verbose("Processed %s initial data ID query rows.", n_rows)
+        # We now recursively populate the data IDs of the rest of the
+        # tree.
+        for branch_dimensions, branch in tree.trunk_branches.items():
+            self.log.debug("Projecting query data IDs to %s.", branch_dimensions)
+            branch.project_data_ids(self.log)
```

Comment:
Maybe make this a method of `_DimensionGroupTree` to improve encapsulation a bit?